### PR TITLE
Added capability to query objects by it's field of type Map for #8231

### PR DESCRIPTION
**Pull request by anatolyD**
Added capability to query objects by it's field of type Map for #8231 
    - `AbstractMultiValueGetter` has new inner state to hold info about the reduceType, not depends on the `int modifier` value anymore, though the value is in sync for backward compatibility.
    - new constant `MODIFIER_NOT_USED` is introduced into `AbstractMultiValueGetter` to inform that the getter refers to the Map kind of reduce
    - `GetterFactory` refactored to avoid duplication of code
    - `GetterFactory` extended with the logic to define return type for non-empty map
    - added test cases for getting Map return type
    - added few basic cases to check query is working for the MapField

Fixes: 
https://github.com/hazelcast/hazelcast/issues/8231
https://github.com/hazelcast/hazelcast/issues/12801

### CODE REVIEW

**Comment by vbekiaris** [comment]
run-lab-run

**Comment by vbekiaris** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/query/impl/getters/AbstractMultiValueGetter.java [hunk 1]
```diff
@@ -63,15 +139,27 @@ Object getValue(Object obj) throws Exception {
...
-            return o;
+        switch (reduceType) {
+            case DO_NOT_REDUCE:
+                return o;
+            case REDUCE_EVERYTHING:
+                MultiResult collector = new MultiResult();
+                reduceInto(collector, o);
+                return collector;
+            case REDUCE_BY_INDEX:
+                return getItemAtPositionOrNull(o, (Integer) modifier);
```
Comment (taburet) [inline hazelcast/src/main/java/com/hazelcast/query/impl/getters/AbstractMultiValueGetter.java]: Casting is not really required here.

**Comment by taburet** [comment]
Good work @anatolyD and привет! :)

It looks like it's possible to support other key types, not just strings, by inferring the key type from non-empty maps and using `AttributeType` to encode this inferred type along with the associated `TypeConverter`. But this is really a big change, so I'm not sure should we try to add this now.

**Comment by mmedenjak** [comment]
Moving to 3.11 as we are releasing 3.10 soon. @anatolyD thank you for your contribution, we will definitely try to merge it. Please address the review comments.

**Comment by anatolyD** [comment]
hi @taburet ! thnx for kind words! :)

 i addressed your findings as well as tried out to use the `AttributeType` thing with the converters. I had to put lazy init of the actual mapKey to be used. But left the String way of the initialization. So even if it is  a non string the single quotes are required. For example: `['42']` for `Integer` key type, instead of `[42]`. Only to have the same way and not to worry about having/non-having quotes for numeric/non-numeric key types...

**Comment by anatolyD** [comment]
@mmedenjak sure. thnx! ;)

**Comment by anatolyD** [comment]
run-lab-run

**Comment by taburet** [comment]
@anatolyD I appreciate your work, but, if I'm not missing something, it should be possible to avoid the single quotes syntax for accessing nested maps. For example, we may infer the key type in `GetterFactory` before the `AbstractMultiValueGetter` construction, as it already done for the value type (`returnType`), and pass it to the `AbstractMultiValueGetter` constructor to facilitate the parsing of the key value. This way we will simplify the parsing, since we can avoid the postponed/lazy parsing, and the single quotes syntax will not be required anymore (TBH, it looks like it's not required even now).

I understand your concerns about empty maps, but `GetterFactory` already returns the special null getters (`NullGetter` and `NullMultiValueGetter`) for them, which are not stored for the reuse (`Getter.isCacheable`), so if we are failed to infer the key type this should not affect `AbstractMultiValueGetter`s.

Probably we may even unify the maps and collections/arrays execution paths in `AbstractMultiValueGetter`, at least partially, in the end collections and arrays may be perceived as maps from integers to objects for our purposes. And thank you for your contribution one more time.

**Comment by anatolyD** [comment]
@taburet interesting. let me dive into it a bit more. originally i introduced quotes to distinguish `any` vs `'string'` types, but as far as we are trying to cover other types, let's see what is doable there.

**Comment by taburet** [comment]
@anatolyD just to make things clear, I was thinking about eliminating the single quotes syntax for types like integers and floats, but there are other tricky cases like, for example, dates or UUIDs. Using them without any quoting will look strange, so probably we should keep the single quotes syntax for now and interpret this as if everything is encoded as a string. @mmedenjak WDYT?

**Comment by Holmistr** [comment]
@taburet Any plans for this PR?

**Comment by taburet** [comment]
@Holmistr The PR must be seriously reworked if we are going to support sub-maps access universally. I think, we may put it on hold for now.

**Comment by anatolyD** [comment]
sorry guys, no time for this one. closing.

**Pull request closed without merging**